File: ax/src/main/java/cc/ifnot/ax/service/ServerService.java
package cc.ifnot.ax.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.os.RemoteException;

import androidx.annotation.Nullable;

import cc.ifnot.ax.IService;
import cc.ifnot.libs.utils.Lg;

/**
 * author: dp
 * created on: 2020/7/18 8:21 AM
 * description:
 */
public class ServerService extends Service {


    @Override
    public void onCreate() {
        super.onCreate();
        Lg.d("onCreate");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Lg.d("onStartCommand");
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        Lg.d("onDestroy");
        super.onDestroy();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        Lg.d("onBind: %s", intent);
        return new IService.Stub() {
            @Override
            public String hello(int i) throws RemoteException {
                Lg.w("hello: %s", i);
                int j = 0;
                do {
                    j += i % 10;
                } while ((i = i / 10) > 0);
                return String.valueOf(j);
            }

            @Override
            public int world() throws RemoteException {
                Lg.w("world");
                return 0xff;
            }
        };
    }

    @Override
    public boolean onUnbind(Intent intent) {
        Lg.d("onUnbind: %s", intent);
        return super.onUnbind(intent);
    }

    @Override
    public void onRebind(Intent intent) {
        Lg.d("onRebind: %s", intent);
        super.onRebind(intent);
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        Lg.d("onTaskRemoved: %s", rootIntent);
        super.onTaskRemoved(rootIntent);
    }
}

File: app/src/main/java/cc/ifnot/app/service/ServerService.kt
package cc.ifnot.app.service

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.os.SystemClock
import cc.ifnot.app.ITest
import cc.ifnot.libs.utils.Lg

/**
 * author: dp
 * created on: 2020/5/29 6:16 PM
 * description:
 */
open class ServerService : Service() {

    companion object {
        const val TAG = "ServerService"
    }

    override fun onBind(intent: Intent?): IBinder? {
        return ServerServiceProxy()
    }

    class ServerServiceProxy : ITest.Stub() {
        override fun iTest0(a: Int) {
            Lg.w("iTest0 start")
            SystemClock.sleep(1000)
            Lg.w("iTest0 end")
        }

        override fun iTest1(a: Int) {
            Lg.w("iTest1 start")
            SystemClock.sleep(1000)
            Lg.w("iTest1 end")

        }

        override fun iTest2(a: Int): Int {
            Lg.w("iTest2 start")
            SystemClock.sleep(1000)
            Lg.w("iTest2 end")
            return a * a
        }

        override fun transact(msg: String?) {
            // do nothing here, but delay
            Lg.w("transact start - %s", msg?.length)
            SystemClock.sleep(1000);
            Lg.w("transact end")
        }

    }
}